### ISSUE TITLE
Return summary objects on list operations 

### ISSUE THREAD

**Issue by tuckerzp** [EasyDynamics/oscal-rest#55]
The full objects are very large, clogging up output and slowing down results. Instead a summary including the metadata would be more useful and the user can then `GET` more specific items.

**Comment by laurelmay** [comment]
This becomes especially notable with really large files, like the SP 800-53 rev5 catalog that includes SP 800-53A. That document is several megabytes in size which can cause some issues.

**Comment by laurelmay** [comment]
I have been thinking about this quite a bit because the performance here is really slowing down a lot of operations. And I think there are two primary means that we can use to support this, each with some pretty significant trade-offs. I would love additional thoughts here as well.

To reiterate the problem space, currently, some Catalogs are 5MB or larger; SSPs and other documents are likely to be very large as well. This introduces performance issues when passing that data over an HTTP response (and when multiple catalogs are included in a response).

For this example, I am going to focus on Catalogs primarily; however, I believe that much of it applies to other document types.

### Summary-style objects

Typically, we would resolve this issue by having a "summary" object that only includes the necessary attributes of the object (with enough to issue a `GET /catalogs/{id}` request); however, there is not a standard representation for these summary-style objects in OSCAL. This introduces several complexities:
- we aren't returning OSCAL data
- implementations of the specification need to implement whatever types we return
- future changes in OSCAL may necessitate changes in this structure

Not returning pure OSCAL isn't a _terrible_ thing but it's a suboptimal precedent and not a habit I want to get into. OSCAL is a well-considered standard and deviating from it is likely to be a red-flag; however, I do think this use-case is one where it may be appropriate. It must be well-considered, though. Different implementations may care about different attributes of a document. Some (like `title`, `version`, or `uuid`) are likely to be needed by nearly everyone but others may care about revision information (or at least the latest revision?). For some documents, actual content may be valuable (consider, it may be useful in the summary to know some data about the hierarchy of imports in profiles).

Implementors needing to care about _our_ types here is important because we want to make sure this REST API specification is easy to adopt. The team at NIST has [`liboscal-java`](https://github.com/usnistgov/liboscal-java) and seems to be working towards implementations in additional languages. This is a great springboard for OSCAL-speaking tooling and if we start adding our own typing on top of this, then implementors have to add implementations of our types too. That's a pain; we could consider providing some implementations (especially handling conversions from other libraries), but even still. Further, through [metaschema](https://github.com/usnistgov/metaschema), XML, JSON, and YAML (and potentially more representations in the future!) are supported in the OSCAL standard. Today, this specification focuses on JSON; however, we are not ruling out considering additional formats as part of the REST API specification in the future. We need to take care that what we create can be represented in these formats (and perhaps should be written in metaschema itself!).

Finally, future changes to OSCAL may introduce complexity here. It's unlikely but possible. If OSCAL were to remove the `title` field (as a wacky example), what would we do if it was included in the summary object?

### Arbitrary queries on documents

Another less-RESTy option that I haven't fully thought out (which is why there's less detail here) would be to allow arbitrary queries on documents -- "give me the title, latest revision data, and UUID of this Catalog". This is really nice to potentially be able to represent; however, it is potentially a lot of complexity to ask of implementations. It's not the end of the world to include something that support XPath or JMESPat but this is a potential barrier. I also really dislike putting just unspecified `object`-type things into an OpenAPI specification but that is likely the best that we can do here.

Beyond pure implementation, this requires more runtime overhead for server implementations. Running these queries isn't free and we'd be shifting a lot of work from the client to the server. One could imagine denial-of-service-style attacks that could come with this (give some ridiculously complex query for a set of known-complex documents) that implementations may not want to support.

### Leave it out of the spec

Any maybe this isn't something we include in the specification! Perhaps, we take a hardline purist OSCAL-only approach as part of the specification on this repository. Individual implementations perhaps can extend this in their own ways but we don't take a stance on anything that isn't in the standard itself. This gives everyone a known predictable baseline but it also introduces a lot of complexity if a client from Implementor A wants to talk to a server from Implementor B. We also have the problem that `GET /catalogs` exists today and is just immensely poorly performing.

### In Summary (😆)

I think that allowing arbitrary queries is the coolest technical solution but perhaps also brings some of the most involved considerations. Leaving it out of the REST API specification is the easiest thing for us to do but makes implementations potentially incompatible. And adding our own document types that derive OSCAL ones seems to introduce some complexity that can be handled by providing additional tooling and libraries but has some longer-term complexity questions.

**Comment by brian-ruf-ezd** [comment]
Addressed in PR #72
